panther_analysis_tool/detection.py

```diff
@@ -37,8 +37,10 @@ class DetectionResult:
     """Class containing the result of running a detection"""
 
     detection_id: str
+    detection_match_alert_value: bool
```

Comment:
Yes, I was struggling with the best approach here becaused `matched` initially made sense to me for policies being false - but when trying to handle "how do I know if I should print anything in `pat`? ... it got messy.  

I like your suggested approach here - simplifies things greatly.

Reply:
> `not self.matcher_alert_value`

should it just default to `False` - i.e. default to non-alert?

Reply:
> What do you think of having two things

That separates the concerns nicely and it's closer to what we have now. Not urgent, but I think we should rename `matched` to `trigger_alert` or something similar, because in the policy context (again due to inverted boolean return value) this may be confusing?

> What this doesn't solve is the issue of "should we run aux functions in this test" (i.e. should this have alerted?).

This is a testing concern, so that's a responsibility of the `testing` module to perform this check based on the `DetectionResult` attributes, since when the detection runs it should have no context about the environment it is running on.

> Would it be 'ok' to print the output of aux function when matched is true or the expected result doesn't match the actual result?

I think what you describe retains the current logic, doesn't it?

I made the following state matrixes to better understand the parameter values in each case and let me know if I'm missing something:


### Policies 

| `matched/trigger_alert` | `detection_output` | `detection_evaluation_failed` |
| --- | --- | --- |
| `True` | `False` | `False` |
| `False` | `None` | `True` |
| `False` | `True` | `False` |

#### Testing


| Should Trigger Alert | `detection_output` | Result |
| --- | --- | --- |
| `False` | `True` | _PASS_ |
| `False` | `False` | _FAIL_ |
| `False` | `None` | _FAIL_  |
| `True` | `None` | _FAIL_  |
| `True` | `False` | _PASS_ |
| `True` | `True` | _FAIL_ |

### Rules

| `matched/trigger_alert` | `detection_output` | `detection_evaluation_failed` |
| --- | --- | --- |
| `True` | `True` | `False` |
| `False` | `None` | `True` |
| `False` | `False` | `False` |

#### Testing

| Should Trigger Alert | `detection_output` | Result |
| --- | --- | --- |
| `False` | `False` | _PASS_ |
| `False` | `True` | _FAIL_ |
| `False` | `None` | _FAIL_  |
| `True` | `None` | _FAIL_  |
| `True` | `True` | _PASS_ |
| `True` | `False` | _FAIL_ |

Reply:
> This is a testing concern, so that's a responsibility of the testing module to perform this check based on the DetectionResult attributes, since when the detection runs it should have no context about the environment it is running on. 

yea, I guess I was just saying it makes the determination within testingpy harder to figure out. 

> I think what you describe retains the current logic, doesn't it?

hmm, I thought that if it expects to not trigger an alert and fails it still wouldn't run the aux function (because they wouldn't have been expected to have run). This is the current check in the the rules-engine:

```
        if self._spec.expectations.rule:
```

re the state diagram: this is super helpful. I think the only thing is rather than `Should Trigger Alert` we get `Expected Return` from the FE.  So we could potentially just make the comparison of the output with the expectation from the FE? (pseudo code):

```
 if expected != detection_output:
     return Fail
# else the expectation matched, so just return whether there was an error
 return error is None
 ```

Reply:
> hmm, I thought that if it expects to not trigger an alert and fails it still wouldn't run the aux function

The `Detection.run` function will still run the aux functions if we have a match we can't avoid this. It's just that in this case the output of these functions is not providing any meaningful feedback to the user, since they shouldn't have been executed, so we can exclude them from the test results.

> So we could potentially just make the comparison of the output with the expectation from the FE? (pseudo code):

I think you can check for error first. Also, if we retain a nullable detection output value, comparing the (always boolean) expectation with the output will always yield correct results: `False is None -> False`